File: src/core/phrase-engine.js
const RARITY_FACTORS = Object.freeze({ common: 1, uncommon: 0.35, rare: 0.06 });
const CONTEXT_KEY_PATTERN = /^[A-Za-z][A-Za-z0-9]*$/;
const KNOWN_PROVIDERS = new Set(['codex', 'claude-code']);

function hasValue(context, key) {
  return Object.prototype.hasOwnProperty.call(context, key) && context[key] !== undefined && context[key] !== null;
}

function isNonNegativeSafeInteger(value) {
  return Number.isSafeInteger(value) && value >= 0;
}

function conditionRule(validate, matches) {
  return Object.freeze({ validate, matches });
}

const CONDITION_RULES = Object.freeze({
  requires: conditionRule(
    (value) => Array.isArray(value)
      && value.length <= 32
      && value.every((key) => typeof key === 'string' && CONTEXT_KEY_PATTERN.test(key)),
    (value, context) => value.every((key) => hasValue(context, key)),
  ),
  batteryEquals: conditionRule(
    (value) => Number.isInteger(value) && value >= 0 && value <= 100,
    (value, context) => context.battery === value,
  ),
  activeCountMin: conditionRule(
    isNonNegativeSafeInteger,
    (value, context) => context.activeCount >= value,
  ),
  remainingMin: conditionRule(
    isNonNegativeSafeInteger,
    (value, context) => context.remaining >= value,
  ),
  remainingEquals: conditionRule(
    isNonNegativeSafeInteger,
    (value, context) => context.remaining === value,
  ),
  durationMinSeconds: conditionRule(
    isNonNegativeSafeInteger,
    (value, context) => context.durationSeconds >= value,
  ),
  lockedMinSeconds: conditionRule(
    isNonNegativeSafeInteger,
    (value, context) => context.lockedSeconds >= value,
  ),
  clickCountMin: conditionRule(
    isNonNegativeSafeInteger,
    (value, context) => context.clickCount >= value,
  ),
  provider: conditionRule(
    (value) => typeof value === 'string' && KNOWN_PROVIDERS.has(value),
    (value, context) => context.provider === value,
  ),
  weekdays: conditionRule(
    (value) => Array.isArray(value)
      && value.length <= 7
      && value.every((weekday) => Number.isInteger(weekday) && weekday >= 0 && weekday <= 6),
    (value, context) => value.includes(context.weekday),
  ),
  hourMin: conditionRule(
    (value) => Number.isInteger(value) && value >= 0 && value <= 23,
    (value, context) => context.hour >= value,
  ),
  hourMax: conditionRule(
    (value) => Number.isInteger(value) && value >= 0 && value <= 23,
    (value, context) => context.hour <= value,
  ),
});

function getConditionValidationError(conditions) {
  if (conditions === undefined) return null;
  if (conditions === null || typeof conditions !== 'object' || Array.isArray(conditions)) {
    return 'conditions must be an object';
  }

  for (const [key, value] of Object.entries(conditions)) {
    if (!Object.prototype.hasOwnProperty.call(CONDITION_RULES, key)) {
      return `an unsupported condition: ${key}`;
    }
    if (!CONDITION_RULES[key].validate(value)) {
      return `an invalid condition: ${key}`;
    }
  }

  if (
    conditions.hourMin !== undefined
    && conditions.hourMax !== undefined
    && conditions.hourMin > conditions.hourMax
  ) {
    return 'an invalid hour range: hourMin cannot exceed hourMax';
  }
  return null;
}

function matchesConditions(conditions = {}, context = {}) {
  if (getConditionValidationError(conditions) !== null) return false;
  return Object.entries(conditions).every(([key, value]) => CONDITION_RULES[key].matches(value, context));
}

function renderTemplate(text, context) {
  let unresolved = false;
  const rendered = text.replace(/\{([A-Za-z][A-Za-z0-9]*)\}/g, (_match, key) => {
    if (!hasValue(context, key)) {
      unresolved = true;
      return '';
    }
    const value = context[key];
    if (!['string', 'number', 'boolean'].includes(typeof value)) {
      unresolved = true;
      return '';
    }
    return String(value);
  });
  return unresolved ? null : rendered;
}

class PhraseEngine {
  constructor(phrases, options = {}) {
    if (!Array.isArray(phrases)) throw new TypeError('PhraseEngine requires a phrase array');
    this.phrases = phrases;
    this.random = options.random || Math.random;
    this.now = options.now || Date.now;
    this.historyLimit = options.historyLimit || 20;
    this.history = [];
    this.lastUsedAt = new Map();
  }

  select(event, context = {}) {
    const now = this.now();
    const candidates = this.phrases
      .filter((phrase) => phrase.event === event)
      .filter((phrase) => matchesConditions(phrase.conditions, context))
      .map((phrase) => ({ phrase, text: renderTemplate(phrase.text, context) }))
      .filter(({ phrase, text }) => {
        if (text === null) return false;
        const lastUsed = this.lastUsedAt.get(phrase.id);
        return lastUsed === undefined || now - lastUsed >= (phrase.cooldownMs || 0);
      });

    if (candidates.length === 0) return null;
    const unseen = candidates.filter(({ phrase }) => !this.history.includes(phrase.id));
    const pool = unseen.length > 0 ? unseen : candidates;
    const weighted = pool.map((entry) => ({
      ...entry,
      effectiveWeight: (entry.phrase.weight || 1) * (RARITY_FACTORS[entry.phrase.rarity || 'common'] || 1),
    }));
    const totalWeight = weighted.reduce((total, entry) => total + entry.effectiveWeight, 0);
    let cursor = Math.min(Math.max(this.random(), 0), 0.999999999999) * totalWeight;
    let chosen = weighted[weighted.length - 1];
    for (const entry of weighted) {
      cursor -= entry.effectiveWeight;
      if (cursor < 0) {
        chosen = entry;
        break;
      }
    }

    this.lastUsedAt.set(chosen.phrase.id, now);
    this.history.push(chosen.phrase.id);
    if (this.history.length > this.historyLimit) this.history.shift();
    return Object.freeze({ ...chosen.phrase, text: chosen.text });
  }
}

module.exports = {
  CONDITION_RULES,
  PhraseEngine,
  getConditionValidationError,
  matchesConditions,
  renderTemplate,
};

File: native-appkit/Sources/BlobfishNative/PhraseEngine.swift
import Foundation

final class PhraseEngine {
    private let phrases: [Phrase]
    private let now: () -> Double
    private let random: () -> Double
    private let historyLimit: Int
    private var history: [String] = []
    private var lastUsedAt: [String: Double] = [:]

    init(
        phrases: [Phrase],
        historyLimit: Int = 20,
        now: @escaping () -> Double = { Date().timeIntervalSince1970 * 1_000 },
        random: @escaping () -> Double = { Double.random(in: 0..<1) }
    ) {
        self.phrases = phrases
        self.historyLimit = historyLimit
        self.now = now
        self.random = random
    }

    func select(event: String, context: [String: JSONValue] = [:]) -> Phrase? {
        let timestamp = now()
        let candidates = phrases.compactMap { phrase -> (Phrase, Double)? in
            guard phrase.event == event,
                  Self.matches(phrase.conditions ?? [:], context: context),
                  let rendered = Self.render(phrase.text, context: context) else { return nil }
            if let previous = lastUsedAt[phrase.id], timestamp - previous < (phrase.cooldownMs ?? 0) { return nil }
            let rarity: Double
            switch phrase.rarity ?? "common" {
            case "uncommon": rarity = 0.35
            case "rare": rarity = 0.06
            default: rarity = 1
            }
            return (Phrase(
                id: phrase.id, event: phrase.event, text: rendered, weight: phrase.weight,
                rarity: phrase.rarity, cooldownMs: phrase.cooldownMs, conditions: phrase.conditions
            ), max(0, phrase.weight ?? 1) * rarity)
        }
        guard !candidates.isEmpty else { return nil }
        let unseen = candidates.filter { !history.contains($0.0.id) }
        let pool = unseen.isEmpty ? candidates : unseen
        let total = pool.reduce(0) { $0 + $1.1 }
        var cursor = min(max(random(), 0), 0.999_999_999) * total
        var selected = pool.last!.0
        for candidate in pool {
            cursor -= candidate.1
            if cursor < 0 { selected = candidate.0; break }
        }
        lastUsedAt[selected.id] = timestamp
        history.append(selected.id)
        if history.count > historyLimit { history.removeFirst(history.count - historyLimit) }
        return selected
    }

    private static func render(_ text: String, context: [String: JSONValue]) -> String? {
        let pattern = try! NSRegularExpression(pattern: #"\{([A-Za-z][A-Za-z0-9]*)\}"#)
        let range = NSRange(text.startIndex..., in: text)
        var rendered = text
        for match in pattern.matches(in: text, range: range).reversed() {
            guard let keyRange = Range(match.range(at: 1), in: text),
                  let wholeRange = Range(match.range(at: 0), in: rendered),
                  let value = context[String(text[keyRange])],
                  let scalar = scalarString(value) else { return nil }
            rendered.replaceSubrange(wholeRange, with: scalar)
        }
        return rendered
    }

    private static func scalarString(_ value: JSONValue) -> String? {
        switch value {
        case .string(let value): return value
        case .number(let value): return value.rounded() == value ? String(Int(value)) : String(value)
        case .bool(let value): return String(value)
        default: return nil
        }
    }

    private static func matches(_ conditions: [String: JSONValue], context: [String: JSONValue]) -> Bool {
        for (key, condition) in conditions {
            switch key {
            case "requires":
                guard let keys = condition.arrayValue?.compactMap(\.stringValue),
                      keys.count == condition.arrayValue?.count,
                      keys.allSatisfy({ context[$0] != nil }) else { return false }
            case "batteryEquals": guard numbersEqual(context["battery"], condition) else { return false }
            case "activeCountMin": guard number(context["activeCount"]) >= number(condition) else { return false }
            case "remainingMin": guard number(context["remaining"]) >= number(condition) else { return false }
            case "remainingEquals": guard numbersEqual(context["remaining"], condition) else { return false }
            case "durationMinSeconds": guard number(context["durationSeconds"]) >= number(condition) else { return false }
            case "lockedMinSeconds": guard number(context["lockedSeconds"]) >= number(condition) else { return false }
            case "clickCountMin": guard number(context["clickCount"]) >= number(condition) else { return false }
            case "provider": guard context["provider"]?.stringValue == condition.stringValue else { return false }
            case "weekdays":
                guard let weekday = context["weekday"]?.numberValue,
                      condition.arrayValue?.contains(where: { $0.numberValue == weekday }) == true else { return false }
            case "hourMin": guard number(context["hour"]) >= number(condition) else { return false }
            case "hourMax": guard number(context["hour"]) <= number(condition) else { return false }
            default: return false
            }
        }
        return true
    }

    private static func number(_ value: JSONValue?) -> Double {
        value?.numberValue ?? -.infinity
    }

    private static func numbersEqual(_ left: JSONValue?, _ right: JSONValue?) -> Bool {
        guard let lhs = left?.numberValue, let rhs = right?.numberValue else { return false }
        return lhs == rhs
    }
}
